sound/soc/sof/pcm.c

```diff
@@ -232,8 +233,18 @@ static int sof_pcm_prepare(struct snd_soc_component *component,
 	if (!spcm)
 		return -EINVAL;
 
-	if (spcm->prepared[substream->stream])
-		return 0;
+	if (spcm->prepared[substream->stream]) {
+		if (!spcm->delayed_stop[substream->stream])
```

Comment:
@plbossart my gut feeling tells that the distortions during xruns are mainly because of not doing just the platform_trigger  during stop to stop the DMA. It seems like a bit of an overkill to do everything else. Also, the flag platform_stop_druing_hw_free was added mainly as a fix for the start/stop tests for IPC4. When switching between pause & running for a pipeline, the host DMA is kept running in the FW, so we must not stop the DMA in the case of start/stop tests. So this patch is likely going to break those tests. And @ujfalusi 's alternate suggestion would work to solve this issue but would definitely break the start/stop tests in CI.

So, the hard part for us is to figure out whether the stop command happened because of a regular stop that will be followed by a hw_free, or a stop that will be followed by a start without a hw_free ot a stop due to an xrun.

Reply:
Thanks for the comments @ranj063 

this patch is supposed to be a 'best-of-both-worlds' solution where we detect an xrun with the stop/prepare pattern. I wanted to have zero impact on the start/stop tests and not break CI - which is the main problem with @ujfalusi's solution, and only change the logic in the xrun case.

So to recap my intent was:
- hw_params/prepare/start/stop/hw_free: no change
- stop/start: no change
- pause/release: no change
-  stop/hw_free: no change
- hw_params/hw_free: no change
- stop/prepare -> CHANGE.

Unfortunately I must have screwed-up something or with my usual luck found a hole in the fw - error 104 is the new 404...